### DISCUSSION TITLE
Berry - working with large native buffers and files

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#18706] [Q&A] [answered]
I can see some file commands for Berry.

If I had a native buffer (jpeg image) and length, is there an easy way to pass a reference to Berry, and then have Berry write the buffer to a file (or do somehting else, like POST it as appended data after settign up the initial HTTP headers in berry)?

**Comment by sfromis** [comment, marked as answer]
The [bytes](https://github.com/berry-lang/berry/wiki/Chapter-7#bytes-class) class is the Berry feature for handling native/binary buffers, including methods like file.readbytes.

**Reply by btsimonh** [reply]
thankyou, again, pointers to the right place (and I'd not found the wiki before!).
To expand on this:
If you have an address and size from TAS, then you can construct a bytes object from them.
(the below is what I learned from the docs, yet to test!).
First you make a `ptr` from the integer address:
```
var addr = tasaddr # an integer address - native buufer address
var len = taslen # an integer length
var p = introspect.toptr(addr) # p is now of type ptr:
```
Then you construct a bytes object from the ptr:
```
var b = bytes(p, len) # b is now an numanaged bytes object:  b.ismapped() should return true
```
as highlighted in the docs, care is required as you can access any memory (including stuff you can't access and cause an access exception).
You also can't free the memory from berry, so there must be some native memory management involved.
Apart from these caveats, it should work like any other bytes object - i.e. we should be able to use it to write files or send HTTP messages.

I will give this a go by creating a TAS command on webcam to get a pointer to a frame, and test by writing a file :).

Edit:  Tested working.  I was able to retrieve a jpeg frame, and write to a file, all with no additional memory allocation from inside Berry.
Hopefully in a week or so, I'll have the webcam updates ready, and scripting of simple image handling available from Berry, as well as motion detect events into Berry :)

**Reply by sfromis** [reply]
While you cannot explicitly free memory via a bytes mapping, any bytes object which was not mapped from some pointer is just a Berry object, available for garbage collection if no longer referenced. Thus, you can do memory management by allocating a buffer in Berry, and pass the pointer to native code if relevant. Obviously, all the normal dangers of memory management applies.
